### ISSUE TITLE
Cast to Chromecast on local network from Jellyfin Player

### ISSUE THREAD

**Issue by spreck** [jellyfin/jellyfin-media-player#490]
**Is your feature request related to a problem? Please describe.**
I have a Jellyfin media server on my local network and can watch media on my phone or computer. I see a cast icon in the top right corner but it says “Google Cast not supported “.

**Describe the solution you'd like**
I’d like to be able to start media on my phone and cast to my TV which has a Chromecast connected to it. Then what should happen is that the media server sends a stream to the Chromecast.
**Describe alternatives you've considered**
I can connect a laptop to the TV directly but this is clunky and it seems unnecessary.

**Additional context**
Again this is all just happening on my local network so ideally all traffic would stay local.

**Comment by DexLuther** [comment]
The cast should be initiated from the phone.

For example, Jellyfin Media Player on Windows 10 says Google Cast unsupported, but I still see my phone. It doesn't matter though unless I wanted to use Jellyfin Media Player (or web) to control my phone (or other device).

Tap the cast button in the Jellyfin phone app and you should see Jellyfin Media Player. Selecting it will let you select media from your phone, which will play on JMP.

**Comment by sevenrats** [comment]
this is not how casting works. the client casts, not the server.

**Comment by temotodochi** [comment]
> this is not how casting works. the client casts, not the server.

Not quite. Media player invokes chromecast to connect to the server to download and stream media. So first your client (android, pc, whatever) needs to see the chromecast device in the same network and the chromecast must be able to connect to the server that actually has the media. This last part is the problem in local area networks because google runs its stuff only on verified public dns servers (8.8.8.8). I'm figuring out myself as well how to do this efficiently that everything stays within my LAN without any public SSL shenanigans if possible. So far it looks like i'd have to get a public SSL certificate that i could then use inside my own LAN with the exact same hostname as i would use publicly. Not optimal.

**Comment by alexandervlpl** [comment]
So how come VLC for Android, `catt` and a hundred other tools can cast directly over the local network no problem, but Jellyfin can't? Clearly there's a way to bypass all the DNS/SSL stuff.

I'm looking at [this guide](https://gist.github.com/Vigrond/1de5fc5ff468a48f053fd455a69c8766) but seems insanely complicated and over-engineered.

**Comment by sevenrats** [comment]
@alexandervlpl it's not clear why you think there is a DNS/SSL issue.

my best understanding is that the chromecast functionality depends on closed source libraries so the web app code switches it on and off depending on whether or not its available.  For example on Android if you install Jellyfin through the playstore, that copy can cast, but if you install the same code with F-Droid, it cannot.  It's just a matter of who Google will permit to use their code, and which code authors are willing to depend on code that with such a questionable license.

**Comment by sevenrats** [comment]
I agree that the guide you linked is pretty crazy. If you are willing to go that far you are much better off using an LSIO-style Kasm-based container to run Chrome inside the networking stack of a jellyfin container. then that Chrome copy could cast right off the local server hardware.  Then, from any platform in existence, you can log into the Kasm web interface to start your streams.

**Comment by alexandervlpl** [comment]
@sevenrats Thanks for the tip, on Android casting does in fact work in the Play version and not the F-Droid version!* Would be great to see this in the docs and in the app description on F-Droid, pretty big limitation. Casting from Chrome on a local PC still seems impossible though, even though video from other sites can cast just fine. I was looking at DNS/SSL because that seems to be what people [try to hack](https://www.reddit.com/r/jellyfin/comments/11pq2wm/jellyfin_connects_to_chromecast_will_not_play/). :man_shrugging: 

*Unrelated, but sadly the Android Jellyfin cast feature seems to be almost unusable with multiple problems and UI glitches. Even VLC on the same phone works much better, and I was hoping to replace it.

**Comment by thornbill** [comment]
Yes the Android app must be installed from the play store or by manually installing the proprietary apk for cast support. The official cast SDKs are all proprietary which is why libre builds cannot include support. We mention this in the project README among other places. I believe we are not able to publish different app descriptions to fdroid and the play store so we are unable to mention this limitation directly in the fdroid listing. (Although if you care enough about FOSS to install from fdroid it seems a bit contradictory to be using Google Cast imho.)

It is my understanding that the VLC project completely reverse engineered the cast protocol to implement their own support which is why they do not have the same limitations. **This is completely out of scope for Jellyfin.**

DNS workarounds for casting are only applicable for casting from a web browser.

I don't know why we are even talking about Android in the repo for the desktop app though...

**Comment by mhalano** [comment]
There is cast support for the Spotify desktop app, that is built with Electron, so gets many Chrome features like Cast support.

**Comment by papadopouloskyriakos** [comment]
Ok then if open source is the issue here, why not an option to cast to "open source speakers" ex. lyrion-lms (squeezelite) / volumio / picoreplayer / etc. ?

**Comment by thornbill** [comment]
Do those speakers use a common API and provide a JavaScript SDK that we could ship?

**Comment by papadopouloskyriakos** [comment]
Thanks for the reply, @thornbill!
Lyrion-LMS has a JSON-RPC API (HTTP, port 9000, /jsonrpc.js) that can control Squeezelite players like piCorePlayer. For example, a POST request like {"id": 1, "method": "slim.request", "params": ["<player-id>", ["playlist", "play", "<stream-url>"]]} plays a track. While there’s no official JavaScript SDK, the API is accessible via standard JS HTTP libraries (e.g., fetch), as seen in projects like Material Skin.
A Jellyfin plugin could use this API to list LMS players in the “Play to” menu and send stream URLs for playback, similar to jellyfin-mpv-shim. Is this something that could fit into Jellyfin’s casting framework? Happy to discuss further!

**Comment by mhalano** [comment]
Maybe we could follow VLC example? It provides a server on the port 8010 that streams to the device.

**Comment by thornbill** [comment]
@llzzrrdd do know if there is documentation to reference? That sounds like it could be doable.

**Comment by papadopouloskyriakos** [comment]
Thanks for the enthusiasm, @thornbill!

The Lyrion-LMS JSON-RPC API is documented at https://lyrion.org/reference/cli/using-the-cli/ and https://lyrion.org/reference/cli/introduction/. It supports commands like playlist play <url> sent via HTTP POST to http://<lms-server>:9000/jsonrpc.js to control Squeezelite players (e.g., piCorePlayer) by player ID. Example: {"id": 1, "method": "slim.request", "params": ["<player-id>", ["playlist", "play", "<stream-url>"]]}. This works with standard JavaScript HTTP libraries (e.g., fetch).

A Jellyfin plugin could list LMS players in the “Play to” menu and send stream URLs for playback, similar to jellyfin-mpv-shim. I’d love to see this feature added for Jellyfin & LMS users like me! Let me know if there’s interest in pursuing it further.

**Comment by andrewrabert** [comment]
This Qt desktop client has been deprecated in favor of a rewrite: https://github.com/jellyfin/jellyfin-desktop
